subscriber_notifee.go

```diff
@@ -0,0 +1,126 @@
+package dht
+
+import (
+	"github.com/jbenet/goprocess"
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/network"
+
+	"github.com/libp2p/go-eventbus"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// subscriberNotifee implements network.Notifee and also manages the subscriber to the event bus. We consume peer
+// identification events to trigger inclusion in the routing table, and we consume Disconnected events to eject peers
+// from it.
+type subscriberNotifee IpfsDHT
+
+func (nn *subscriberNotifee) DHT() *IpfsDHT {
+	return (*IpfsDHT)(nn)
+}
+
+func (nn *subscriberNotifee) subscribe(proc goprocess.Process) {
+	dht := nn.DHT()
+
+	dht.host.Network().Notify(nn)
+	defer dht.host.Network().StopNotify(nn)
+
+	var err error
+	evts := []interface{}{
+		&event.EvtPeerIdentificationCompleted{},
+	}
+
+	// subscribe to the EvtPeerIdentificationCompleted event which notifies us every time a peer successfully completes identification
+	sub, err := dht.host.EventBus().Subscribe(evts, eventbus.BufSize(256))
```

Comment:
We should subscribe _then_ go async.

Reply:
Otherwise, we'll miss peers.

Reply:
Makes sense. Are we concerned at all that people might start the DHT after they're connected to a peer and therefore not receive event? Do we need to go through the peerstore to look for DHT peers after we've subscribed to the eventbus to make sure we don't miss peers?

Reply:
Hm. Well, you're right. We _could_ subscribe async and go through the set of connected peers immediately after subscribing. That's actually the _most_ robust solution.

Reply:
Given https://github.com/libp2p/go-libp2p-kad-dht/pull/469#discussion_r387437195 should we try and start the eventbus outside of the goroutine or does the panic seem reasonable?

Reply:
Given that, the panic sounds reasonable regardless. In terms of where to start: either works but walking over all connected peers is actually the best solution so doing this outside the goroutine isn't really necessary.